File: projects/m5/001-budget-app/solutions/LucaBertoldi/budget.php
<?php 


class Categories  {
    public $name;
    public $ledger = [];
    public $totalAmount;

    public function __construct($name) {
        $this->name = $name;
    }

    public function __toString() {
        $output_string = str_pad("{$this->name}", 40, "*", STR_PAD_BOTH);
        foreach($this->ledger as $value) {
            $decoded_array = json_decode($value);
            $output_string .= 
            "<br/>" . 
            "<div style='width:300px; position:relative'>" 
            . substr($decoded_array->description,0,23) . 
            "<span style='text-align:right; position:absolute; right:0'>" 
            . substr(round($decoded_array->amount, 2),0,8) . 
            "</span></div>";
        };
        return $output_string;
    }

    public function deposit($amount, $description = "") {
        $this->totalAmount = $amount;
        $depositResult = array("amount" => $amount, "description" => $description);
        array_push($this->ledger, json_encode($depositResult));
    }

    public function withdraw($amount, $description = "") {
        if($this->check_funds($amount)) {
            $this->totalAmount -= $amount;
            $depositResult = array("amount" => -$amount, "description" => $description);
            array_push($this->ledger, json_encode($depositResult));
            return true;
        } else {
            return false;
        }
    }

    public function get_balance() {
        return $this->totalAmount;
    }

    public function transfer ($amount, $budget_category) {
        if($this->check_funds($amount)) {
            $this->withdraw($amount, "Trasfer to `{$budget_category->name}`");
            $budget_category->deposit($amount, "Trasfer from `{$this->name}`");
            return true;
        } else {
            return false;
        }
        
    }

    protected function check_funds($amount) {
        if($this->totalAmount < $amount) {
            return false;
        } else {
            return true;
        }
    }
}



function create_spend_chart($categories) {
    $output_string = "<br><br>Percentage spent by category<br><table>";
    $percentage = [];
    foreach($categories as $category) {
        $percentage[$category->name] = findPercentage($category);
    }
    for($i=100; $i>=-10; $i-=10) {

        if($i === -10) {
            $output_string .= "<tr style='text-align:right'><td></td>";
        } else {
            $output_string .= "<tr style='text-align:right'><td>" . $i  . "|</td>";
        }
        
        
        foreach($percentage as $key => $value) {            
            $output_string = categoryNames($value, $i, $output_string, $key);
        }
        
    }         
    $output_string .= "</table>";

    return $output_string;
}

function findPercentage($category) {
    $sum = 0;
    $category_array = $category->ledger;
    foreach($category_array as $value) {
        $decoded_array = json_decode($value);
        if($decoded_array->amount < 0) {
            $sum += $decoded_array->amount;
        } else {
            $total = $decoded_array->amount;
        }
    };
    return (abs($sum / $total) * 100);
}

function categoryNames($value, $i, $output_string, $key) {
    if($value >= $i) {
        if($i == -10) {
            $output_string .= "<td style='text-align:left; writing-mode: vertical-lr;text-orientation: upright; border-top: 2px dashed black;'>" . $key . "</td>";
        } else {
            $output_string .= "<td style='text-align:right'>0</td>";
        }
       
    } else {
        $output_string .= "<td style='text-align:right'></td>";
    }

    return $output_string;
}
?>

File: projects/m5/001-budget-app/solutions/RoccoFigliamonti/budget.py
class Category:

    def __init__(self,name):
        self.name = name
        self.ledger = []

    def __str__(self):
        statement = str(self.name).center(30, "*") + "\n"
        for el in self.ledger:
            statement += el["description"][:23].ljust(23)
            statement += ("{0:.2f}".format(el['amount'])).rjust(7) + "\n"
        statement += "Total: " + str(self.get_balance()) + "\n"
        return statement

    
    def deposit(self, amount, description=""):
        deposit_dict = {"amount":amount, "description": description}
        self.ledger.append(deposit_dict)

    def get_balance(self):
        amount_list = []
        for el in self.ledger:
            amount_list.append(el["amount"])
        return sum(amount_list)

    def check_funds(self,amount):
        if amount > self.get_balance():
            return False
        else:
            return True
    
    def withdraw(self, amount, description=""):
        if self.check_funds(amount):
            withdraw_dict = {"amount":-amount, "description": description}
            self.ledger.append(withdraw_dict)
            return True
        else:
            return False
        
    def transfer(self, amount, category):
        if self.check_funds(amount):
            self.withdraw(amount, f"Transfer to {category.name}")
            category.deposit(amount, f"Transfer from {self.name}")
            return True
        else:
            return False



def create_spend_chart(categories):
    #creating a dictionary key = category, value = total amount of withdraw for that category
    expenses_dict = {}
    for category in categories:
        category_exp = 0
        for el in category.ledger:
            if el["amount"] < 0 and ("Transfer") not in el['description']: #if it's a withdraw
                category_exp += abs(el["amount"])
            expenses_dict[category.name] = category_exp 
    #new dict with relative values in %
    tot_expenses = sum(expenses_dict.values())
    expen_perc = {k: round(v/tot_expenses*100) for k, v in expenses_dict.items()}

    #creating chart %
    chart = "Percentage spent by category\n"
    for i in range(100,-10, -10):
        chart += str(i).ljust(3) + "|"
        for el in categories:
            if expen_perc[el.name]>= i:
                chart += "o "
            else:
                chart += ""
        chart += "\n"

    #chat label ???
    return chart


